src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1849,38 +1849,97 @@ protected override void AfterLeftChildHasBeenVisited(BoundBinaryOperator binary)
             ResultType = InferResultNullability(binary, leftType, rightType);
 
             BinaryOperatorKind op = binary.OperatorKind.Operator();
+
+            // learn from non-null constant
+            BoundExpression operandComparedToNonNull = null;
+            if (skipImplicitNullableConversions(binary.Left).ConstantValue?.IsNull == false)
+            {
+                operandComparedToNonNull = binary.Right;
+            }
+            else if (skipImplicitNullableConversions(binary.Right).ConstantValue?.IsNull == false)
+            {
+                operandComparedToNonNull = binary.Left;
+            }
+
+            if (operandComparedToNonNull != null)
+            {
+                switch (op)
+                {
+                    case BinaryOperatorKind.Equal:
+                    case BinaryOperatorKind.GreaterThan:
+                    case BinaryOperatorKind.LessThan:
+                    case BinaryOperatorKind.GreaterThanOrEqual:
+                    case BinaryOperatorKind.LessThanOrEqual:
+                        operandComparedToNonNull = SkipReferenceConversions(operandComparedToNonNull);
+                        splitAndLearnFromNonNullTest(operandComparedToNonNull, caseToLearn: true);
+                        return;
+                    default:
```

Comment:
We should learn from a `!=` test as well.  In `if (o != 1)` we know `o` is not null in the `else` branch.

Reply:
Please either make a change or file an issue (your choice)

---
In reply to: [263641782](https://github.com/dotnet/roslyn/pull/33929#discussion_r263641782) [](ancestors = 263641782)